src/CMakeLists.txt

```diff
@@ -121,6 +121,65 @@ add_test(NAME rust-unit-tests
+# * When we make the shim no_std, we'll want all dependencies built with
+#   only the features that the shim asks for, so that another crate
+#   doesn't end up enabling a feature that brings in `std`.
+
+# We need to set the soname of the shim for the injector lib to link to it correctly.
+# See https://github.com/rust-lang/rust/pull/31170.
+# We can't just set it in the shim's [build].rustflags, because that gets ignored
+# completely if the RUSTFLAGS env variable is set.
+# https://doc.rust-lang.org/cargo/reference/config.html#buildrustflags
+set(RUSTFLAGS "$RUSTFLAGS -C link_args=-Wl,-soname,libshadow_shim.so")
+set(CARGO_ENV_VARS "")
+set(CARGO_ENV_VARS "${CARGO_ENV_VARS} RUSTFLAGS=\"${RUSTFLAGS}\"")
+set(CARGO_ENV_VARS "${CARGO_ENV_VARS} CFLAGS=\"${RUST_CFLAGS}\"")
+set(CARGO_ENV_VARS "${CARGO_ENV_VARS} CXXFLAGS=\"${RUST_CXXFLAGS}\"")
+set(CARGO_ENV_VARS "${CARGO_ENV_VARS} CC=\"${CMAKE_C_COMPILER}\"")
+set(CARGO_ENV_VARS "${CARGO_ENV_VARS} CXX=\"${CMAKE_CXX_COMPILER}\"")
+# Use a separate targetdir so that this and "main" rust build
+# don't invalidate eachother's build caches.
+# TODO: I'm not sure *why* they invalidate each-other's build caches.
+# Theoretically cargo is supposed to avoid that sort of issue by mangling
+# the compiler artifacts with the appropriate metadata to be able to keep
+# multiple configurations around at once.
+set(TARGETDIR "${CMAKE_CURRENT_BINARY_DIR}/shim/target")
+ExternalProject_Add(
+    rust-shadow-shim-project
+    PREFIX ${CMAKE_CURRENT_BINARY_DIR}
+    BINARY_DIR ${CMAKE_CURRENT_SOURCE_DIR}
+    BUILD_ALWAYS 1
+    DOWNLOAD_COMMAND ""
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND bash -c " \
+        ${CARGO_ENV_VARS} \
+        cargo build \
+        -p shadow-shim \
+        ${RUST_TARGET_FLAG} \
+        ${RUST_BUILD_FLAG} \
+        --target-dir \"${TARGETDIR}\" \
+        --features \"${RUST_FEATURES}\" \
+        "
+    BUILD_BYPRODUCTS
+      ${TARGETDIR}/${RUST_TARGET}/debug/libshadow_shim.so ${TARGETDIR}/${RUST_TARGET}/release/libshadow_shim.so
+    INSTALL_COMMAND ""
+    LOG_BUILD OFF
+)
+add_library(shadow-shim SHARED IMPORTED GLOBAL)
+add_dependencies(shadow-shim rust-shadow-shim-project)
+set_target_properties(shadow-shim PROPERTIES IMPORTED_LOCATION_DEBUG ${TARGETDIR}/${RUST_TARGET}/debug/libshadow_shim.so)
+set_target_properties(shadow-shim PROPERTIES IMPORTED_LOCATION_RELEASE ${TARGETDIR}/${RUST_TARGET}/release/libshadow_shim.so)
+# Use this instead when we're able to bump the cmake min version to >= 3.21:
+# > install(IMPORTED_RUNTIME_ARTIFACTS shadow-shim LIBRARY DESTINATION lib)
+install(FILES ${TARGETDIR}/${RUST_TARGET}/debug/libshadow_shim.so DESTINATION lib CONFIGURATIONS Debug)
+install(FILES ${TARGETDIR}/${RUST_TARGET}/release/libshadow_shim.so DESTINATION lib CONFIGURATIONS Release)
+# Rust's std needs -ldl and -lpthread. Remove this when the shim is no_std.
+target_link_libraries(shadow-shim INTERFACE dl pthread)
```

Comment:
Why does cmake even do any linking, shouldn't cargo handle that?

Reply:
libinjector (a C target) links against the shim. Thinking about it some more though, I *think* we don't need this since the `.so` already has its runtime dependencies embedded in its metadata. Removed.